remake dopedxpy-sc-fermi

alexsquires opened the pull request:
- [x] rename bulk_vasprun, not clear that this needs to be for a high quality dos
- [x] use _format_defect_name() in examples (is this still a private function? If it's externally useful, should it be changed?).
- [x] remove ..._and_save() from FermiSolver methods
- [x] ~~defect_levels argument when annealing becomes a boolean~~
- [x] docstrings
- [x] multiprocessings
- [x] tests
- [x] suppress noisy warnings
- [x] tutorials
- [x] grid-searching
- [x] effective dopant response
- [x] "exceptions" for FermiSolverPyScFermi

kavanase commented:
Looking good @alexsquires. I like the mix-in/abstract class approach.

The `_interpolate_chempots()` function is quite nice, and I think will likely be one of the mostly-used features from these additions. From what we were discussing before in the `#doped` Slack, would it be possible to also integrate a grid-scanning approach too? As in, the current function linearly interpolates between two points (chem pot limits), which is nice and v useful in many cases, but if for instance we have a 3D chemical potential space, our min/max of the property of interest could also lie somewhere in the middle of this space, rather than along any linear path between two vertices/limits.

Ofc the user could iterate over all possible vertex/limit combinations if they knew what they were doing, which would be a good stab at this and get you most of the way there I imagine (kind of like a 'band structure path' approach), but still is only covering certain 'high-symmetry' paths in chem pot space. So if a grid approach was also possible (where e.g. the user can just set the chempot spacing in eV, or total number of grid points), that would also be v nice I think – should be doable with some of the `scipy` grid space interpolation functions I think? And/or `pymatgen` chemical potential diagram methods maybe? 
If it was possible to implement relatively painlessly, could we add these two options? ☝️ 🙏 

Sorry to be piling on the requests, but I guess the other main use case I'd imagine for this part of the code would be for the more complex defect thermodynamics analysis that `py-sc-fermi` allows, where you want to fix certain defect/species etc concentrations and perform some constrained solutions. Is it possible to integrate this to the `fermi_solver` code?

Also just fyi, about your earlier `_format_defect_name()` question, yes it is actually quite externally useful and I'll make it a public function in the next (minor) release.

alexsquires commented:
Yeah, dw this is all on the list, should be ready to go by the end of the
week, I sort of trashed everything to start again (for about the 6th time
now), and just adding everything back in piecewise to make sure it's all
consistent with the latest thermodynamics object.

On Thu, 15 Feb 2024 at 04:58, Seán Kavanagh ***@***.***>
wrote:

> Looking good @alexsquires <https://github.com/alexsquires>. I like the
> mix-in/abstract class approach.
>
> The _interpolate_chempots() function is quite nice, and I think will
> likely be one of the mostly-used features from these additions. From what
> we were discussing before in the #doped Slack, would it be possible to
> also integrate a grid-scanning approach too? As in, the current function
> linearly interpolates between two points (chem pot limits), which is nice
> and v useful in many cases, but if for instance we have a 3D chemical
> potential space, our min/max of the property of interest could also lie
> somewhere in the middle of this space, rather than along any linear path
> between two vertices/limits.
>
> Ofc the user could iterate over all possible vertex/limit combinations if
> they knew what they were doing, which would be a good stab at this and get
> you most of the way there I imagine (kind of like a 'band structure path'
> approach), but still is only covering certain 'high-symmetry' paths in chem
> pot space. So if a grid approach was also possible (where e.g. the user can
> just set the chempot spacing in eV, or total number of grid points), that
> would also be v nice I think – should be doable with some of the scipy
> grid space interpolation functions I think? And/or pymatgen chemical
> potential diagram methods maybe?
> If it was possible to implement relatively painlessly, could we add these
> two options? ☝️ 🙏
>
> Sorry to be piling on the requests, but I guess the other main use case
> I'd imagine for this part of the code would be for the more complex defect
> thermodynamics analysis that py-sc-fermi allows, where you want to fix
> certain defect/species etc concentrations and perform some constrained
> solutions. Is it possible to integrate this to the fermi_solver code?
>
> Also just fyi, about your earlier _format_defect_name() question, yes it
> is actually quite externally useful and I'll make it a public function in
> the next (minor) release.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SMTG-Bham/doped/pull/46#issuecomment-1945954259>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFMK74XGZEUFF3TPF7MZHH3YTXZ5JAVCNFSM6AAAAABBVDPFSSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNBVHE2TIMRVHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

alexsquires commented:
@adair-nicolson , could you re-parse your Cu2SiSe3 data now that everything is relatively final and I can add the grid searching stuff back in?

alexsquires commented:
@kavanase proposing this as the full functionality - do you want to take a look at the example ipynb while I finish off the tests before I get to comfortable?

kavanase commented:
@alexsquires sorry for the delay in getting to this! Busy time...

Looking very very nice, thanks very much for your work with this! 🙌 

Some small queries from me:
- Can `multiplicity_scaling` be handled internally by `FermiSolverPyScFermi` (but still retaining the option to set by the user), by just dividing the volume of the bulk supercell in `defect_thermodynamics` by that of `bulk_dos_vr` (and maybe a quick warning if this isn't close to an integer)? Would make it easier and remove a potential source of error if so.
```python
py_fs = FermiSolverPyScFermi(defect_thermodynamics=thermodynamics, bulk_dos_vr=vasprun_path, multiplicity_scaling=32)
```
- I updated the `exceptions` example to add one that matches better the example given in the text (i.e. mobile interstitials), to avoid confusion with this and give a relevant example. Can `exceptions` be reworded to something a bit more intuitive, like `non_frozen` or `equilibrium_defects` or something like that?
- Some of the public functions (like `scan_dopant_concentration`, `interpolate_chemical_potentials` etc) don't have docstrings. Could you add these when you get a chance please? (Noticed when I was figuring out how `exceptions` worked, useful to have a quick help message for how parameters like this work)
- When I went to commit, I got a load of `pre-commit` formatting warnings. Could you install `pre-commit` (running `pre-commit install` in the `doped` directory) to handle the remaining issues please? (Dealt with some here, but not all)

A general question: (will also look into the code in more detail myself)
- How are metastable defects handled in `FermiSolverPyScFermi`? Is it equivalent to the way they're handled in the `DefectThermodynamics` functions (at equilibrium and upon cooling)? Might possibly be worth quickly checking, by setting e.g. the V_Cd^0 metastable state to only be 0.01 eV higher energy than the ground-state (to amplify the metastable effects), and see if the `doped` and `FermiSolverPyScFermi` results are consistent or differ in this case?

kavanase commented:
Fyi, the `py-sc-fermi` interface tutorial is now accessible here: https://doped.readthedocs.io/en/dopey_fermi/py_sc_fermi_interface_tutorial.html

kavanase commented:
@alexsquires just to flag, it seems like there are some unused files included here in the CdTe examples directory. Before merging can we check that only the ones needed are included, to avoid bloating the repo?
Also just fyi, I removed `CdTe_Thesis_pyscfermi.ipynb`, `data/dos_vasprun.xml` and `vasprun_nsp.xml` here as they weren't being used, but if they were intended to be used for something else then please revert!

kavanase commented:
Hi @alexsquires!
Some notes on these changes from going through in more depth:

- I added the `effective_dopant_concentration` parameter to the `doped` `DefectThermodynamics` functions as discussed (and verified it matches with `py-sc-fermi`). Also in the updated notebook, the effective dopant concentration is now plotted alongside, which I think nicely shows it's effects on the carrier/defect concentrations (ofc you can tell this anyway from reading off the axes, but nice to see it directly plotted I think)
<img width="759" alt="image" src="https://github.com/SMTG-Bham/doped/assets/51478689/b8e98ec6-fb6c-4e35-9e29-3f7176e671ad">

 - In doing this, I noticed a bug in the `FermiSolverPyScFermi` code for `effective_dopant_concentration` with annealing & quenching. Previously this was giving flat concentrations for the native defects:
![image](https://github.com/SMTG-Bham/doped/assets/51478689/a9b06117-164b-4bae-bcf2-0742131f0a61)
       I realised this was related to the placement of the code which adds the effective dopant concentration in `generate_annealed_defect_system`. Fixed now, but also returns a stream of `RuntimeWarning`s, would you be able to update to ignore these (for the case where `FermiSolverPyScFermi` is being used for this)? Just not sure where/how you're handling warnings like this!
<img width="717" alt="image" src="https://github.com/SMTG-Bham/doped/assets/51478689/c8a63152-af5a-408b-9253-3a967fff1ac9">

- As a suggestion (open to being challenged on this!), I refactored `fix_defect_species` to (`not`) `fix_charge_states` as I think this is a bit more intuitive in general? I think in the field "species" is used both to mean defect type + charge, or just defect type (regardless of whether one is technically more correct), so best to rename to something like this?
- Can we add to `effective_dopant_concentration` docstrings a note that negative means acceptor doping and positive means donor doping? Just there's no actual mention of this in the docstrings so could be confusing for the user, and isn't explicitly stated in the tutorial from what I can see. Also would be good to explicitly state that `effective_dopant_concentration` is in units of cm^-3
- Would also be good to note in the docstrings and notebook the different chemical potential definitions / usages for `doped`/`py-sc-fermi` (formal vs DFT values)
- Lastly, I tried using `fs.scan_chemical_potential_grid()` with `FermiSolverDoped` for the last bit with Cu2SiSe3 – as this should also work fine with the `doped` solver right? But it crashed out with a `ValueError`, which I think is probably related to how the chemical potentials are being fed to `doped` here (i.e. with this difference between how they should be handled for `doped` vs `py-sc-fermi`). I left this in the notebook, would you mind having a quick look at how this is integrated and if it's an easy fix?

kavanase commented:
@alexsquires are you still working away on this or do you want me to look over? Just checking!

alexsquires commented:
Mostly there, just checks and warning suppression to go

On Sat, 6 Apr 2024, 19:51 Seán Kavanagh, ***@***.***> wrote:

> @alexsquires <https://github.com/alexsquires> are you still working away
> on this or do you want me to look over? Just checking!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SMTG-Bham/doped/pull/46#issuecomment-2041166010>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFMK74SFPG6DLBBD3HS5QRTY4A73PAVCNFSM6AAAAABBVDPFSSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANBRGE3DMMBRGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

alexsquires commented:
@kavanase has something changed with the chemical potentials? Cu2SiSe3 example no longer has limits wrt the elements, CdTe does?

kavanase commented:
@alexsquires I'm not totally sure what you're referring to here. I checked the chempots json files in the examples folders (`examples/CdTe` & `examples/Cu2SiSe3`), and they both seem to have the `"limits_wrt_el_refs"` subdict present:

CdTe:
<img width="608" alt="image" src="https://github.com/SMTG-Bham/doped/assets/51478689/e4aa19b6-2d5e-41af-8979-de671c27c643">

Cu2SiSe3:
![image](https://github.com/SMTG-Bham/doped/assets/51478689/b2a167ae-dbf7-4838-b0ba-a6c5fa665252)

alexsquires commented:
For making the `fermi_solver` chempot interface consistent with `doped.thermodynamics`, do you mind if the chempot handling methods in thermodynamics are no longer private methods?

kavanase commented:
Yeah sure! Not an issue at all

alexsquires commented:
I've made sure all the chempots labelling is consistent and added the option to specify chemical potentials as a limit, rather than just a dictionary. It's a bit different from the DefectThermodynamics object, because I've made doped-formatted chempots a required attribute of the FermiSolvers. Any comments before I finalise it?

kavanase commented:
@alexsquires this looks really great, thanks for pushing forward with it!
I've had time to go through the code now, and made some small fixes on the way.

Some last comments from me:
- [ ] Just curious, is there a reason why `joblib.Parallel` is used rather than `multiprocessing`?
- [ ] Somewhat related; have you done any profiling/speed tests to see how much this speeds up these calculations? Would be curious to see! (On my to-do list anyway)
- [ ] `processes=1` by default, so no multiprocessing. Is there a particular reason for this? Is it worth having the default number of `processes` set to >1? (there are functions that tell you the number of CPUs available, like `cpu_count()` used elsewhere in the code). Could do something like the implementations in other parts of `doped`, where it's set to higher numbers depending on the expected computational cost (and thus potential speedup)?
- [ ] For the chempot scanning functions, setting the dependent variable is mostly unnecessary/inconsequential right? Like the output is basically the same, but it might just affect the grid placement somewhat, as they're just linearly related? So best to have this as an optional argument, and if not set just pick the last element in the formula or something like that?
e.g.:
![image](https://github.com/user-attachments/assets/1ce1c139-fc58-456b-93e1-1186d118c0b1)
- [ ] On a related point, I don't totally get what `n_points` is exactly, and it isn't really explained by the docstrings. E.g. when I set it to 50 for a 2D grid, it returns much more than 50x50 = 2500 datapoints (but then also in the output plot it looks like less than this many datapoints?). Would be good to clarify this in the docstrings (and maybe tutorial) of how this sets the grid to be scanned over.
![image](https://github.com/user-attachments/assets/de3e5453-bf8f-4ece-a886-7395f2625b54)
> n_points (int): number of points to scan
> n_points (int): The number of points to generate along each axis of the grid.

- [ ] **Bug:** For the `min_max_X` function, which is a v nice feature here, I think there's a slight issue. I noticed that the output chemical potential in the Cu2SiSe3 example in the tutorial (-0.41 eV for Cu) is not the lowest Cu chemical potential (-0.46 eV) as it should be, as this gives the highest hole concentration as shown in the plots just above it. So it outputs $p$<sub>max</sub> as 7.43e19 when it's actually 9.44e19 (so the difference is much larger than the relative tolerance of 0.001). Looking into the code and adding some debug info, I think the reason is related to the fact that the new sub-grid generated by the loop is set to the midpoint of the current vertices and the 'target chempot', but if the target chempot is at the border / on a vertex then this creates a new grid which is actually further away from the target? This can also be seen by the fact that the target value actually gets worse rather than better with the second loop in this example case. Would you be able to look at fixing this please? (Hopefully easy 🙏 )

![image](https://github.com/user-attachments/assets/6df610fd-5b25-4021-bbb1-012dc972e801)

- [ ] I'll also help out with this, but would be good to flesh out the tests a bit. Maybe some integration tests with the CdTe/MgO/Cu2SiSe3/Sb2Si2Te6 example systems would be a quick and easy way to test most functionality?


Some minor points:
- [ ] I guess this was on the finalisation to-do list, but just to flag, would be good to write/flesh out the docstrings for the different functions. 
- [ ] Do we need to add a note in the docstrings/tutorial at the end to point out that `n_points` may need to be lowered for more complex systems (i.e. more chempot dimensions)?
- [ ] Would be good to explicitly list what functionality these methods offer that users can't get with the standard `DefectThermodynamics` functions, I left a space blank in the tutorial for this:
<img width="776" alt="image" src="https://github.com/user-attachments/assets/21100a94-9a6e-4051-8a97-75ead20d4bc0">


Not including it in the tutorial as it's irrelevant, but nice to plot the chempot scanning results here alongside the plot in the SI of Adair's paper:
![image](https://github.com/user-attachments/assets/26c7206f-71b5-49f3-8acd-30c2fae2c0db)

alexsquires commented:
> Just curious, is there a reason why joblib.Parallel is used rather than multiprocessing?

Annoyingly, I don't remember the specifics. Serves me right for letting this drag on so long. It's to do with passing defect system objects to `multiprocessing`. I played around with different ways to do it and this is the least ugly I could come up with.

> Somewhat related; have you done any profiling/speed tests to see how much this speeds up these calculations? Would be curious to see! (On my to-do list anyway) 

Not extensively. Re this and the next point, I noticed that the defaults in doped were processes `!= None`. I was leaving this hanging until some profiling was done. Once you get down to it on your to-do list, maybe we can come up with a sensible setting? 

> For the chempot scanning functions, setting the dependent variable is mostly unnecessary/inconsequential right? Like the output is basically the same, but it might just affect the grid placement somewhat, as they're just linearly related? So best to have this as an optional argument, and if not set just pick the last element in the formula or something like that? 

Just binned it, good call. 

> On a related point, I don't totally get what n_points is exactly, and it isn't really explained by the docstrings. E.g. when I set it to 50 for a 2D grid, it returns much more than 50x50 = 2500 datapoints (but then also in the output plot it looks like less than this many datapoints?). Would be good to clarify this in the docstrings (and maybe tutorial) of how this sets the grid to be scanned over.

`n_points` is the number of points generated along each chemical potential axis on a regular grid. each point is then assessed as being either inside or outside the hull. So the number of points on the plot will always be less than `npoints * npoints`. The reason why the results dataframe is so long is because these are long format (because it makes plotting, espesh with non matplotlib (seaborn) packages, easier. And long-form data is generally more machine readable. So its n_points (within the hull) * the number of defect species (+ electrons and holes). I will make sure this is better reflected in the docs. 

> min_max_X bug. 

I think I fixed this by always including the vertices of the grid in any search, rather than the in-hull points in the interpolation. Which is just a good thing to have anyway. 

Related to this, and the tests, do you have chempot data for the non CdTe+Cu2SiSe3 examples? Any help with testing is gratefully received, and then I can make sure the new search works for all the systems.

I will work on the minor points in the meantime.

kavanase commented:
Ok cool! 
Yeah with `multiprocessing` the shared memory handling / 'stateful' calculations / pickling can be annoying issues. Makes sense.

And yep for `processes` the defaults in `doped` depend on the calculation task. Typically tries to set `processes` to `num_cpus - 1` when the workload is estimated to be sufficiently substantial that multiprocessing will give a significant speedup.

All sounds good!

For the chempots Q, yes there are also `chempots` for V<sub>2</sub>O<sub>5</sub>, MgO and Sb<sub>2</sub>O<sub>5</sub> (see here: https://github.com/SMTG-Bham/doped/blob/main/tests/test_thermodynamics.py#L159), along with their `DefectThermodynamics` objects, but all binaries... I think there might have been an example chempots file for YTOS previously, but not anymore. Will search for this. Failing that, if you need it for a more complicated system, could ask Jiayi for her LNMO data? (Checked the publications and don't think it was shared online). Or could just use the MP data to generate some test chempots for a YTOS/others? 
Or if useful, using your LiNiO2 data?

kavanase commented:
There is this YTOS phase diagram object in the git history (though not actually sure if it's parsed hybrid data or just taken from the Materials Project). Adair was following up that defects project before, so should have the parsed chempots for it?
(https://github.com/SMTG-Bham/doped/blob/2.3.0/examples/competing_phases/ytos_phase_diagram.json)

alexsquires commented:
Docstrings and typing updated to be doped consistent, and example notebook now populated with some papers that people can go and look at for.

kavanase commented:
@alexsquires going through now. One question, would it make sense to allow initialisation of `FermiSolverDoped/PyScFermi` objects using `FermiSolver(..., backend="doped/py-sc-fermi")` to make it a bit cleaner for the user? 
Maybe defaulting to `py-sc-fermi` if installed, otherwise `doped`. If you agree, I can implement as it should just be a small update

alexsquires commented:
Such a good idea that I'm embarrassed I didn't think of it before

kavanase commented:
Ok cool will implement! Also to check, is there any point keeping `multiplicity_scaling` as an optional input? Like the default handling (with a minor update to allow some negligible numerical differences) should basically always be the desired usage?

```python
        if multiplicity_scaling is None:
            ms = self.defect_thermodynamics.defect_entries[0].defect.structure.volume / self.volume
            # check multiplicity scaling is almost an integer
            if not np.isclose(ms, round(ms)):
                raise ValueError(
                    "The multiplicity scaling factor is not almost an integer. "
                    "Please specify a multiplicity scaling factor."
                )
            self.multiplicity_scaling = round(ms)
        else:
            self.multiplicity_scaling = multiplicity_scaling
```

kavanase commented:
Also did you check with the latest code that the tutorial all works fine?

alexsquires commented:
> Ok cool will implement! Also to check, is there any point keeping `multiplicity_scaling` as an optional input? Like the default handling (with a minor update to allow some negligible numerical differences) should basically always be the desired usage?
> 
> ```python
>         if multiplicity_scaling is None:
>             ms = self.defect_thermodynamics.defect_entries[0].defect.structure.volume / self.volume
>             # check multiplicity scaling is almost an integer
>             if not np.isclose(ms, round(ms)):
>                 raise ValueError(
>                     "The multiplicity scaling factor is not almost an integer. "
>                     "Please specify a multiplicity scaling factor."
>                 )
>             self.multiplicity_scaling = round(ms)
>         else:
>             self.multiplicity_scaling = multiplicity_scaling
> ```

Yeah, we can get rid of it. I suppose it is a hang over from my very "what if?" approach to py-sc-fermi where most of the fun was just playing around with the defect system and seeing how everthing changed, but if people want that, they can go to py-sc-fermi 🤓 

On the thing of having fermisolver with a back-end option, I think the default should just be doped in all cases. This is doped after all, and it is marginally faster for the thing I reckon 90% of users will want, which is the convenience functions with quench+anneal. 

Everything was working locally when I pushed, but I was in a bit of a frenzy, have you found something that doesn't work?

kavanase commented:
Ok! And yeah I guess if it's an advanced enough user that they were playing around with things like that, they could also achieve the same effects other ways (e.g. by just adding an entry to `DefectEntry.degeneracy_factors` etc).

For using the `doped` backend as default, ok makes sense! (Especially if it's currently a bit quicker). I think it should be fairly easy to add a convenience method that converts from the `doped` backend to the `py-sc-fermi` one, so might try add this so that the `doped` backend can be used by default, and then switches to the `py-sc-fermi` one if needed (e.g. for `free_defects`) if it's installed. I guess the `free_defects` behaviour could also be added to the `doped` Fermi solving functions too, which in theory should be fairly straightforward.

For the last bit about something that doesn't work, it was just that I moved this `_handle_chempots()` line a couple lines up (because the `chempots` output was un-used so I assumed it was meant to be before when `chempots` is used), but this broke some things, so the line is just removed now and all seems to be working as expected.
https://github.com/SMTG-Bham/doped/blob/b8cf01515a0e3e64fb4899b9d986634696b68626/doped/interface/fermi_solver.py#L1451

kavanase commented:
Added some of the changes mentioned above now.

I played around with `joblib`/`multiprocessing` for this, with some profiling (though not so easy). It seems that thread locking and synchronization is causing the bottlenecks and slow speeds once attempting parallelization, which I think implies the need for shared resources but this shouldn't be the case afaik... Anyway it's pretty fast now as is. If someone needs to run it faster they can ask about multiprocessing or try themselves

Question: For `scan_chempots`, it's meant to be a list of chemical potentials that are then scanned over yeah? I guess we expect this to be a rare use case, with `interpolate_chempots` or `scan_chemical_potential_grid` usually preferred by the users. Either way, the current code is broken as it doesn't loop over the input list. Just checking there's still a use case for this function before I go trying to fix?

A benefit of the merge to `FermiSolver` with backend choice now is that we also minimise docstring duplication for the methods, so easier to maintain and update

kavanase commented:
Ok some notes on the above. I updated the `chempots`/`limit`/`el_refs` inputs to be flexible and adopt a similar format to that in `thermodynamics`, to make things easier for the user, for maintenance etc.

I also (mostly, at least) fixed the `scan_chempots()` function mentioned above. 

From playing around, I realised that `ChemicalPotentialGrid` and associated functions seems to only work for ternary compounds? e.g. with CdTe:
![image](https://github.com/user-attachments/assets/cc8cbecf-8b1d-4337-af7e-bffdf89fd9e2)
<img width="770" alt="image" src="https://github.com/user-attachments/assets/d7eac2df-b7f8-42cb-9888-e55c59afaf35">

and Na2FePO4F:
<img width="760" alt="image" src="https://github.com/user-attachments/assets/ab56df00-5e30-4312-9df1-7cca7a449407">
<img width="776" alt="image" src="https://github.com/user-attachments/assets/cf1ed2b7-e023-47fd-b04e-3bfa7c77a13d">
(I've now added the `get_doped_chempots_from_entries` convenience function to `chemical_potentials` so can generate example `chempots` from the Materials Project quickly and easily this way, as shown).

Would you be able to look at this, so that it works for all dimensions?

Also for the tests, it would be most efficient to use a decorator (like [this](https://github.com/SMTG-Bham/doped/blob/develop/tests/test_chemical_potentials.py#L68) for testing both the new and legacy MP APIs in the chemical potentials tests) which tests each analysis function for both the `doped` and `py-sc-fermi` backends, rather than duplicating code. Would you mind adding this, and some tests for the `scan_...` functions (ideally with a variety of binary/ternary/quaternary+ systems)? 🙏

kavanase commented:
@alexsquires, I was looking again at the current code in this PR (hopefully nearly ready to go 🤞) as I'm currently trying to use it in some of my work (on selenium defects). I was hoping to use the `py-sc-fermi` use case where one can set a fixed concentration of a defect/impurity species, and solving with this constraint, but I think this is not implemented as far as I can tell?

I checked back and this was mentioned way back at the start of this PR, and the `effective_dopant_concentration` option was then added, but I think this other functionality wasn't?
> Sorry to be piling on the requests, but I guess the other main use case I'd imagine for this part of the code would be for the more complex defect thermodynamics analysis that py-sc-fermi allows, where you want to fix certain defect/species etc concentrations and perform some constrained solutions. Is it possible to integrate this to the fermi_solver code?

Just checking I'm not missing something / avoiding redundant work. I can add this to the code if it's not there.

alexsquires commented:
> @alexsquires, I was looking again at the current code in this PR (hopefully nearly ready to go 🤞) as I'm currently trying to use it in some of my work (on selenium defects). I was hoping to use the `py-sc-fermi` use case where one can set a fixed concentration of a defect/impurity species, and solving with this constraint, but I think this is not implemented as far as I can tell?
> 
> I checked back and this was mentioned way back at the start of this PR, and the `effective_dopant_concentration` option was then added, but I think this other functionality wasn't?
> 
> > Sorry to be piling on the requests, but I guess the other main use case I'd imagine for this part of the code would be for the more complex defect thermodynamics analysis that py-sc-fermi allows, where you want to fix certain defect/species etc concentrations and perform some constrained solutions. Is it possible to integrate this to the fermi_solver code?
> 
> Just checking I'm not missing something / avoiding redundant work. I can add this to the code if it's not there.

I think it would be nice to maybe have it accept a dictionary of defects and concentraitons, and optionally a charge state so e.g. 

```
fixed_concentrations = {"v_Cd": x, "v_Te_+1" : y}
```

This could fairly easily be an additional kwarg of any of the `py-sc-fermi` methods?


> @alexsquires something I was also thinking about. Do you think it still makes sense to have this under the separate `interface` module? Of course `py-sc-fermi` is one of the optional backends here, but most of the code is backend-agnostic. What about moving it into `doped.thermodynamics`? I don't have any strong feelings about this, just thinking if it's the better category fit

I also thinks this makes sense. I will fix everything up at the beginning of next week, then we could consider moving it?

kavanase commented:
Sounds good!

alexsquires commented:
Do you have a local version of the example notebook that is not pushed? There are some issues running the version on GH? I can fix them if not, just checking I'm not reinventing the wheel

kavanase commented:
As far as I can tell, my local version isn't different to the pushed version, but in case there's any formatting/etc issues, this is my local version, as well as another notebook I was using for some quick tests before.
[Archive.zip](https://github.com/user-attachments/files/17016856/Archive.zip)

alexsquires commented:
This covers the bug features and feature requests. Do you want to consider the merge into thermodynamics?

kavanase commented:
Ok great! Yes let's merge to `doped.thermodynamics` – and for the `ChemicalPotentialGrid` code, does it make sense to put it in `doped.chemical_potentials`?

alexsquires commented:
Migration finished! Just updating the tests.

kavanase commented:
@alexsquires trying to blast through this and get it merged and released asap.
I'll still have a few to-dos after (like supporting per-charge outputs as in the `DefectThermodynamics` methods, rather than always grouped for each defect, if possible). Currently trying to beef up the tests a bit, and test actual numbers to catch cases where it's not working as expected. 
One issue I've hit is with `effective_dopant_concentration` with the `py-sc-fermi` backend. I'm not sure if it's an issue with the `FermiSolver(backed="py-sc-fermi")` code in `doped`, or `py-sc-fermi` – would you be able to have a look at this?

```
/Users/kavanase/miniconda3/bin/python /Applications/PyCharm.app/Contents/plugins/python-ce/helpers/pycharm/_jb_pytest_runner.py --target test_fermisolver.py::TestFermiSolverWithLoadedData.test_pseudo_equilibrium_solve_doped_backend -- --last-failed -vv 
Testing started at 17:58 ...
Launching pytest with arguments --last-failed -vv test_fermisolver.py::TestFermiSolverWithLoadedData::test_pseudo_equilibrium_solve_doped_backend --no-header --no-summary -q in /Users/kavanase/Packages/doped/tests

============================= test session starts ==============================
collecting ... collected 1 item
run-last-failure: rerun previous 1 failure

test_fermisolver.py::TestFermiSolverWithLoadedData::test_pseudo_equilibrium_solve_doped_backend 

======================== 1 failed, 1 warning in 12.67s =========================
FAILED [100%]Testing with doped backend
Testing with py-sc-fermi backend

tests/test_fermisolver.py:552 (TestFermiSolverWithLoadedData.test_pseudo_equilibrium_solve_doped_backend)
self = <test_fermisolver.TestFermiSolverWithLoadedData testMethod=test_pseudo_equilibrium_solve_doped_backend>
backend = 'py-sc-fermi'

    @parameterize_backend()
    def test_pseudo_equilibrium_solve_doped_backend(self, backend):
        """
        Test ``pseudo_equilibrium_solve`` method for both backends.
        """
        solver = self.solver_doped if backend == "doped" else self.solver_py_sc_fermi
        single_chempot_dict, el_refs = solver._get_single_chempot_dict(limit="Te-rich")
    
        # Call the method
>       concentrations = solver.pseudo_equilibrium_solve(
            annealing_temperature=800,
            single_chempot_dict=single_chempot_dict,
            el_refs=el_refs,
            quenched_temperature=300,
            effective_dopant_concentration=1e16,
            append_chempots=True,
        )

test_fermisolver.py:562: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../doped/thermodynamics.py:5047: in pseudo_equilibrium_solve
    conc_dict = defect_system.concentration_dict()
../../../miniconda3/lib/python3.10/site-packages/py_sc_fermi/defect_system.py:380: in concentration_dict
    e_fermi = self.get_sc_fermi()[0]
../../../miniconda3/lib/python3.10/site-packages/py_sc_fermi/defect_system.py:221: in get_sc_fermi
    q_tot = self.q_tot(e_fermi=e_fermi)
../../../miniconda3/lib/python3.10/site-packages/py_sc_fermi/defect_system.py:329: in q_tot
    lhs_def, rhs_def = self.total_defect_charge_contributions(e_fermi)
../../../miniconda3/lib/python3.10/site-packages/py_sc_fermi/defect_system.py:307: in total_defect_charge_contributions
    [
../../../miniconda3/lib/python3.10/site-packages/py_sc_fermi/defect_system.py:308: in <listcomp>
    ds.defect_charge_contributions(e_fermi, self.temperature)
../../../miniconda3/lib/python3.10/site-packages/py_sc_fermi/defect_species.py:416: in defect_charge_contributions
    for q, concd in self.charge_state_concentrations(e_fermi, temperature).items():
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = 
Dopant, nsites=1
fixed [c] = 7.004017649987961e-07
  q=+1.0, [c]=7.004017649987961e-07, deg=1

e_fermi = -0.7575499999999993, temperature = 300

    def charge_state_concentrations(
        self, e_fermi: float, temperature: float
    ) -> Dict[int, float]:
        """at a given Fermi energy and temperature, calculate the concentrations
        of the different ``DefectChargeStates`` of this ``DefectSpecies``.
    
        Args:
            e_fermi (float): Fermi energy
            temperature (float): temperature
    
        Returns:
            Dict[int, float]: key-value pairs of charge of each
            ``DefectChargeState`` and the concentration of the
            ``DefectChargeState`` with that charge, i.e.
            {``DefectChargeState.charge``: concentration}
        """
    
        var_concs = self.variable_conc_charge_states()
        fixed_concs = self.fixed_conc_charge_states()
    
        cs_concentrations = {
            cs.charge: cs.get_concentration(e_fermi, temperature) * self.nsites
            for cs in var_concs.values()
        }
        for q, cs in fixed_concs.items():
            cs_concentrations[q] = cs.get_concentration(e_fermi, temperature)
    
        if self.fixed_concentration is not None:
            fixed_conc_chg_states = sum(
                [
                    c
                    for q, c in cs_concentrations.items()
                    if q in self.fixed_conc_charge_states()
                ]
            )
            variable_conc_chg_states = sum(
                [
                    c
                    for q, c in cs_concentrations.items()
                    if q in self.variable_conc_charge_states()
                ]
            )
            constrained_conc = self.fixed_concentration - fixed_conc_chg_states
>           scaling = constrained_conc / variable_conc_chg_states
E           ZeroDivisionError: float division by zero

../../../miniconda3/lib/python3.10/site-packages/py_sc_fermi/defect_species.py:391: ZeroDivisionError

Process finished with exit code 1
```

alexsquires commented:
It came to me in a dream.

When the “Dopant” was designated as a fixed-concentration `DefectSpecies` as part of the "annealing", all its associated `DefectChargeState` objects were also set to fixed concentrations. During the `py-sc-fermi` calculations, the code attempts to handle variable versus fixed charge states by determining the relative ratios of each `DefectChargeState` for a given defect species. However, if every charge state within that species is fixed, there are no variable states to compare against—leading to an edge case in the solver. I've fixed this in the logic for generating annealed defect systems and I will add a note to the `py-sc-fermi` codebase to catch scenarios where an entire defect species is fixed, so that similar cases don’t cause confusion in the future.

kavanase commented:
Great, thanks @alexsquires. Will try push through on updating the other tests

kavanase commented:
@alexsquires I'm aiming to merge this and release this weekend / early this week.

I renamed the `min_max_X` method to `optimise` as I think this might be a bit more intuitive / obvious as a user-facing function. For the tutorial, I think we can probably remove the `doped`/`py-sc-fermi` comparisons; useful to have for testing (I have added some `TODO` notes to the tests to use these as handy integration tests, by generating these plots and testing with `pytest-mpl`), but I think it's not needed in the tutorials and will just confuse users. Any thoughts on this or all other changes lmk!

Some notes/thoughts just for reference:
- Per-charge outputs are currently not supported for `FermiSolver`, but are for `DefectThermodynamics`, and so only total defect concentrations are accessible. This is useful information to users in a lot of cases, so would be good to include in future. I think this is possible with `py-sc-fermi`, just needs to use the `decomposed` option with `concentration_dict`?
- I think I'm still not sure how metastable defects (two defects of the same type and charge state) are handled in `py-sc-fermi`'s functions. They're handled natively in `doped` with the appropriate Boltzmann summation (from what I remember before). Anyway I guess this is unlikely to matter significantly in most cases.
- Future TODO: Should also allow just specifying an extrinsic element for ``fixed_defects``,  to allow the user to specify the known concentration of a dopant / over/under-stoichiometry of a given element (but with unknown relative populations of different possible defects) -- realistically the most commonly desired option (but would require a bit of refactoring from the current `py-sc-fermi` implementation).
    - The `DefectThermodynamics` JSONs in the repo for extrinsic-doped Selenium ([link](https://chemrxiv.org/engage/chemrxiv/article-details/6706a93612ff75c3a1f8c995)) would be a good test case for this.
- Future TODO: In future the ``fixed_defects``, ``free_defects`` and ``fix_charge_states`` options may be added to the ``doped`` backend (in theory very simple to add)
- Future TODO: Add per-site option like in ``doped`` ``DefectThermodynamics``, should be quick to add
- Currently the `fixed_defects`, `free_defects` and `fix_charge_states` options aren't really tested, I've added a TODO note for this.
- I have some examples now in the tests where the extremum position for a defect/carrier concentration doesn't occur at a limiting chemical potential (e.g. CdTe from my thesis, V_S in Sb2S3 just about (https://pubs.acs.org/doi/10.1021/acsenergylett.4c02722)), which I think will be good to show in the tutorial as this is the main case where this function is particularly powerful.
- It will also be good to use the `scan_X` functions now in the main [thermodynamics tutorial](https://doped.readthedocs.io/en/latest/thermodynamics_tutorial.html) as this should now be the most convenient and recommended way of doing this, unless extra control is needed e.g. to do the bandgap scissoring I showed for CdTe.

I'll try get to these when I can find some time (lol), but if you had a chance at any point to try tackle some of these would be much appreciated!

alexsquires commented:
Sounds good. I should be able to work on some of these this week.

```
Future TODO: Should also allow just specifying an extrinsic element for
fixed_defects, to allow the user to specify the known concentration of a
dopant / over/under-stoichiometry of a given element (but with unknown
relative populations of different possible defects) -- realistically the
most commonly desired option (but would require a bit of refactoring from
the current py-sc-fermi implementation).

   - The DefectThermodynamics JSONs in the repo for extrinsic-doped
   Selenium (link
   <https://chemrxiv.org/engage/chemrxiv/article-details/6706a93612ff75c3a1f8c995>)
   would be a good test case for this

```
Never thought of this, nice idea. I was talking to Ben about some plans for
some new py-sc-fermi functionality and I can add this to the pile
(including proper metastable defects and site competition). Just to note,
as well as `per_charge` information in the concentration dict, you can get
at `per_site` concentrations via setting `per_volume` to false. Just in
case you get there first.







On Sat, 18 Jan 2025 at 20:06, Seán Kavanagh ***@***.***>
wrote:

> @alexsquires <https://github.com/alexsquires> I'm aiming to merge this
> and release this weekend / early this week.
>
> I renamed the min_max_X method to optimise as I think this might be a bit
> more intuitive / obvious as a user-facing function. For the tutorial, I
> think we can probably remove the doped/py-sc-fermi comparisons; useful to
> have for testing (I have added some TODO notes to the tests to use these
> as handy integration tests, by generating these plots and testing with
> pytest-mpl), but I think it's not needed in the tutorials and will just
> confuse users. Any thoughts on this or all other changes lmk!
>
> Some notes/thoughts just for reference:
>
>    - Per-charge outputs are currently not supported for FermiSolver, but
>    are for DefectThermodynamics, and so only total defect concentrations
>    are accessible. This is useful information to users in a lot of cases, so
>    would be good to include in future. I think this is possible with
>    py-sc-fermi, just needs to use the decomposed option with
>    concentration_dict?
>    - I think I'm still not sure how metastable defects (two defects of
>    the same type and charge state) are handled in py-sc-fermi's
>    functions. They're handled natively in doped with the appropriate
>    Boltzmann summation (from what I remember before). Anyway I guess this is
>    unlikely to matter significantly in most cases.
>    - Future TODO: Should also allow just specifying an extrinsic element
>    for fixed_defects, to allow the user to specify the known
>    concentration of a dopant / over/under-stoichiometry of a given element
>    (but with unknown relative populations of different possible defects) --
>    realistically the most commonly desired option (but would require a bit of
>    refactoring from the current py-sc-fermi implementation).
>       - The DefectThermodynamics JSONs in the repo for extrinsic-doped
>       Selenium (link
>       <https://chemrxiv.org/engage/chemrxiv/article-details/6706a93612ff75c3a1f8c995>)
>       would be a good test case for this.
>    - Future TODO: In future the fixed_defects, free_defects and
>    fix_charge_states options may be added to the doped backend (in theory
>    very simple to add)
>    - Future TODO: Add per-site option like in doped DefectThermodynamics,
>    should be quick to add
>    - Currently the fixed_defects, free_defects and fix_charge_states
>    options aren't really tested, I've added a TODO note for this.
>    - I have some examples now in the tests where the extremum position
>    for a defect/carrier concentration doesn't occur at a limiting chemical
>    potential (e.g. CdTe from my thesis, V_S in Sb2S3 just about (
>    https://pubs.acs.org/doi/10.1021/acsenergylett.4c02722)), which I
>    think will be good to show in the tutorial as this is the main case where
>    this function is particularly powerful.
>    - It will also be good to use the scan_X functions now in the main thermodynamics
>    tutorial
>    <https://doped.readthedocs.io/en/latest/thermodynamics_tutorial.html>
>    as this should now be the most convenient and recommended way of doing
>    this, unless extra control is needed e.g. to do the bandgap scissoring I
>    showed for CdTe.
>
> I'll try get to these when I can find some time (lol), but if you had a
> chance at any point to try tackle some of these would be much appreciated!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/SMTG-Bham/doped/pull/46#issuecomment-2599860926>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFMK74UZFSK2YF56RKG2XZL2LKX4DAVCNFSM6AAAAABBVDPFSSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOJZHA3DAOJSGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kavanase commented:
Ok cool. All `FermiSolver` tests passing now, merging this to `develop` (and hopefully releasing soon!).
Thanks again for this addition @alexsquires!
